npm/metrics/prometheus-metrics.go

```diff
@@ -53,71 +39,183 @@ const (
-		numIPSetEntries = createGauge(numIPSetEntriesName, numIPSetEntriesHelp, false)
-		ipsetInventory = createGaugeVec(ipsetInventoryName, ipsetInventoryHelp, false, setNameLabel, setHashLabel)
+	// TODO introduce isFanOut parameter to determine when to create fan-out controller/daemon metrics
+	if haveInitialized {
+		klog.Infof("metrics have already been initialized")
+	} else {
+		initializeDaemonMetrics()
+		initializeControllerMetrics()
+		// TODO include dataplane health metrics:
+		// num failures for apply ipsets, updating policies, deleting policies, and running periodic policy tasks, etc.
 		log.Logf("Finished initializing all Prometheus metrics")
 		haveInitialized = true
 	}
 }
 
-// ReinitializeAll creates/replaces Prometheus metrics. This function is intended for UTs.
-// Be sure to reset helper variables e.g. ipsetInventoryMap.
+// ReinitializeAll creates/replaces Prometheus metrics.
+// This function is intended for UTs.
 func ReinitializeAll() {
+	klog.Infof("reinitializing Prometheus metrics. This may cause error messages of the form: 'error creating metric' from trying to re-register each metric")
 	haveInitialized = false
 	InitializeAll()
-	ipsetInventoryMap = make(map[string]int)
 }
 
 // GetHandler returns the HTTP handler for the metrics endpoint
-func GetHandler(isNodeLevel bool) http.Handler {
-	return promhttp.HandlerFor(getRegistry(isNodeLevel), promhttp.HandlerOpts{})
+func GetHandler(registryType RegistryType) http.Handler {
+	if !haveInitialized {
+		// not sure if this will ever happen, but just in case
+		klog.Infof("in GetHandler, metrics weren't initialized. Initializing now")
+		InitializeAll()
+	}
+	return promhttp.HandlerFor(getRegistry(registryType), promhttp.HandlerOpts{})
 }
 
-func register(collector prometheus.Collector, name string, isNodeLevel bool) {
-	err := getRegistry(isNodeLevel).Register(collector)
+// initializeDaemonMetrics creates non-controller metrics
+func initializeDaemonMetrics() {
+	// CLUSTER METRICS
+	numACLRules = createClusterGauge(numACLRulesName, numACLRulesHelp)
+	numIPSets = createClusterGauge(numIPSetsName, numIPSetsHelp)
+	numIPSetEntries = createClusterGauge(numIPSetEntriesName, numIPSetEntriesHelp)
+	ipsetInventory = createClusterGaugeVec(ipsetInventoryName, ipsetInventoryHelp, ipsetInventoryLabels)
+	ipsetInventoryMap = make(map[string]int)
+
+	// NODE METRICS
+	addACLRuleExecTime = createNodeSummary(addACLRuleExecTimeName, addACLRuleExecTimeHelp)
+	addIPSetExecTime = createNodeSummary(addIPSetExecTimeName, addIPSetExecTimeHelp)
+}
+
+// initializeControllerMetrics creates metrics modified by the controller
+func initializeControllerMetrics() {
+	// CLUSTER METRICS
+	numPolicies = createClusterGauge(numPoliciesName, numPoliciesHelp)
+	// TODO include health metrics: num failures for validating policies & ipsets
+
+	// NODE METRICS
+	addPolicyExecTime = createNodeSummaryVec(addPolicyExecTimeName, "", addPolicyExecTimeHelp, addPolicyExecTimeLabels)
```

Comment:
addPolicyExecTime should be a daemon metric, like ACLRuleExectime or aaddIPsetExectime.

Reply:
i understand that it is being used in controller space now, but ideally this should be taken care of in dp.AddPOlicy